### PR TITLE
Feat/dark themes

### PR DESCRIPTION
**Pull request by m0ksem**
firefox_a82WxMzvMX
Better themes support and dark themes as well.

# Breaking changes.
- Colors changed according to https://github.com/epicmaxco/vuestic-ui/issues/744#issuecomment-1117201636 (removed white, dark, black and gray colors)
Added new colors to color-config: textLight, textDark, background, backgroundSoft, backgroundMute, shadow, focus.
- Added auto text color for VaTimePicker and VaDatePicker relative to their backgrounds. Here can be performance issues, but it is optimized by maximum.
- Removed useless CSS variables
- Some component colors maybe changed to fit new color pallet.

## TODO
- [ ] Better types for vuestic color config colors (we must require colors such as primary, background etc, but user must be able to add own.)
- [ ] Check if all colors used correctly. 

closes https://github.com/epicmaxco/vuestic-ui/issues/744
closes https://github.com/epicmaxco/vuestic-ui/issues/1590
closes https://github.com/epicmaxco/vuestic-ui/issues/1255
closes https://github.com/epicmaxco/vuestic-ui/issues/1385
closes https://github.com/epicmaxco/vuestic-ui/issues/757
closes https://github.com/epicmaxco/vuestic-ui/issues/1858
closes https://github.com/epicmaxco/vuestic-ui/issues/1867

### CODE REVIEW

**Comment by m0ksem** [comment]
Theme demk from docs for now.

image
image

**Comment by xx13** [comment]
@m0ksem thank you, here are my suggestions and questions:

- `primary` -> `accent` as this colour is mostly used for interactive elements that should stand out of plain text
- we can consider adding `accent-2` if the colour theme is more diverse, and an extra colour can be used to add style to static elements, like bullet points, icons, etc); a good example might be the first version of Vuestic where we had blue and green as accent colours
- `textDark` -> `primary` as it is the most commonly met style on the screen, and it can be applied not only to text, but also to other elements, like icons (if there is only dark theme, then textDark most likely will be white)
- `textLight` -> `inverted / tretiary` that is most likely to be #fff if `primary` is set to ~black, which can be used for elements over dark background to have accessible contrast

if we are to introduce a dark theme, it might have sense to introduce something like `primary const` and `inverted / tretiary const` that keeps the colour value even when switching from light to dark themes. a good example is when you have a white playback button over a video, and it should remain white for both light and dark themes.

I'd have background styles as following:
- `background` -> `background-primary` (the most commonly met style for widgets, cards, or the entire screen if there are no cards/widgets)
- `background-soft` -> `background-secondary`(background for inner groups, secondary elements, notes)
- `background-muted`->`background-form-element`(input bg, inline select, etc)
- if we are speaking of admin themes, I think having separate values like `background-sidebar`, `background-topbar`, `background-screen` will make customization more intuitive (these variables may have the same values as other `background-…` vars)

do you think it makes sense to consider adding `extra` colours, that would match the theme, and will be used for tags, graphs, etc? or will those colour be defined in a different way?

**Comment by xx13** [comment]
`misc form` can be renamed to `divider` as it's used for dividers and for some variants of inputs, so `divider` is its primary function

**Comment by xx13** [comment]
I'll post colour hex value updates to this thread

**Comment by m0ksem** [comment]
@xx13, for text, we calculate color relative to background color. So on white background we use darkColor, on dark - whiteColor.
So user don't need to change darkColor to '#fff' in dark theme. The logic here that darkColor stay dark in any theme.

**Comment by m0ksem** [comment]
> @m0ksem thank you, here are my suggestions and questions:
> 
>     * `primary` -> `accent` as this colour is mostly used for interactive elements that should stand out of plain text
> 
>     * we can consider adding `accent-2` if the colour theme is more diverse, and an extra colour can be used to add style to static elements, like bullet points, icons, etc); a good example might be the first version of Vuestic where we had blue and green as accent colours
> 
>     * `textDark` -> `primary` as it is the most commonly met style on the screen, and it can be applied not only to text, but also to other elements, like icons (if there is only dark theme, then textDark most likely will be white)
> 
>     * `textLight` -> `inverted / tretiary` that is most likely to be #fff if `primary` is set to ~black, which can be used for elements over dark background to have accessible contrast
> 
> 
> if we are to introduce a dark theme, it might have sense to introduce something like `primary const` and `inverted / tretiary const` that keeps the colour value even when switching from light to dark themes. a good example is when you have a white playback button over a video, and it should remain white for both light and dark themes.
> 
> I'd have background styles as following:
> 
>     * `background` -> `background-primary` (the most commonly met style for widgets, cards, or the entire screen if there are no cards/widgets)
> 
>     * `background-soft` -> `background-secondary`(background for inner groups, secondary elements, notes)
> 
>     * `background-muted`->`background-form-element`(input bg, inline select, etc)
> 
>     * if we are speaking of admin themes, I think having separate values like `background-sidebar`, `background-topbar`, `background-screen` will make customization more intuitive (these variables may have the same values as other `background-…` vars)
> 
> 
> do you think it makes sense to consider adding `extra` colours, that would match the theme, and will be used for tags, graphs, etc? or will those colour be defined in a different way?

- `accent` is a good replacement for `primary`, but having `accent-2` is kind of strange solution. Probably user will seek for `accent-1` or event `accent-0`. `primary` and `secondary` easy to understand, because it is two opposites.
- since we calculate text color relative to it background `text-dark` and `text-light` must stay constant.
- how about `backround-form`? Just simply to remember across a user's app.
- user can change any component color via props or global config, so we don't need extra color for one type of component. For example, carousel buttons use '#00000aa' color, but it is absent in config, because this color is used only in VaCarousel component.
For admin template, we can define any amount of colors we want. Here we just need a couple of them, so the framework can work.



In Android apps it is common to have the following colors: primary, secondary, triadic, complementary We can actually rename misc-form to triadic color. Same with backgorunds.

File: packages/ui/src/composables/useElementRef.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+type Maybe<T> = T | undefined
+export type MaybeElement<T extends Maybe<HTMLElement>> = T | ComponentPublicInstance
+
+const unrefElement = <T extends Maybe<HTMLElement>>(el: Ref<MaybeElement<T>>): T => {
+  const e = unref(el)
+  return (e as ComponentPublicInstance)?.$el ?? e
+}
+
+export const useElementRef = <T extends Maybe<HTMLElement>>() => {
+  const el = ref<MaybeElement<T>>()
```
Comment (RVitaly1978) [inline packages/ui/src/composables/useElementRef.ts:12]:
```suggestion
  const el = shallowRef<MaybeElement<T>>()
```

File: packages/ui/src/composables/useElementTextColor.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * Returns text color based on element background.
+ *
+ * If HTMLElement provided will take element background counting opacity.
+ *
+ * If string (CSS color) provided will use it as background color.
+ *
+ * If element not provided current instance element will be used.
+ */
+export const useElementTextColor = (background?: Ref<HTMLElement | CssColor | undefined> | CssColor) => {
+  if (isRefString(background)) {
```
Comment (RVitaly1978) [inline packages/ui/src/composables/useElementTextColor.ts:18]: what if `background` type === `CssColor`?

**Comment by m0ksem** [comment]
image
image

**Comment by m0ksem** [comment]
Merging to prevent conflicts in feture PRs. Stuff to do:

- [ ] Optimize useElementBackgorund (with caching)
- [ ] Update colors demo.
- [ ] Setup theme service config.
- [ ] Maybe rename color names.

**Pull request merged**